webapp/src/main/java/com/box/l10n/mojito/service/smartling/ThirdPartyTextUnitMatchingService.java

```diff
@@ -0,0 +1,201 @@
+import com.box.l10n.mojito.smartling.response.FilesResponse;
+import com.box.l10n.mojito.smartling.response.SourceStringsResponse;
+import com.box.l10n.mojito.utils.Optionals;
+import com.box.l10n.mojito.utils.PageFetcherSplitIterator;
+import org.slf4j.Logger;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+
+import java.time.Instant;
+import java.util.*;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import java.util.stream.StreamSupport;
+
+import static com.box.l10n.mojito.utils.Predicates.logIfFalse;
+import static org.slf4j.LoggerFactory.getLogger;
+
+/**
+ * Service to manage matching between mojito text units and third party text units
+ **/
+@Service
+public class ThirdPartyTextUnitMatchingService {
+    /**
+     * logger
+     */
+    private static Logger logger = getLogger(ThirdPartyTextUnitMatchingService.class);
+
+    @Autowired
+    ThirdPartyTextUnitRepository thirdPartyTextUnitRepository;
+
+    @Autowired
+    SmartlingClient smartlingClient;
+
+    @Autowired
+    ThirdPartyTextUnitSearchService thirdPartyTextUnitSearchService;
+
+    String delimiter = "#@#";
+
+    private String fileRegex = ".*\\d+_(singular|plural)_source\\.xml";
+
+    private String pluralFileIndicator = "plural";
+
+    void getSmartlingThirdPartyTextUnits(String projectId) {
+        FilesResponse files = smartlingClient.getFiles(projectId);
+        for (File file : files.getResponse().getData().getItems()) {
+            String fileUri = file.getFileUri();
+            if (fileUri.matches(fileRegex)) {
+                processFile(fileUri, projectId);
+            }
+        }
+    }
+
+    public void processFile(String fileUri, String projectId) {
+        int maxResults = 500;
+
+        // pull all strings for a file then match
+        PageFetcherSplitIterator<StringInfo> sourceStringsPageFetcherSplitIterator = new PageFetcherSplitIterator<>(
+            (offset, limit) -> {
+                List<StringInfo> stringInfoList = new ArrayList<>();
+                SourceStringsResponse sourceStringsResponse = smartlingClient.getSourceStrings(projectId, fileUri, offset);
+                if (sourceStringsResponse.isSuccessResponse()) {
+                    stringInfoList = sourceStringsResponse.getResponse().getData().getItems();
+                }
+                return stringInfoList;
+            }, maxResults
+        );
+        Stream<StringInfo> stream = StreamSupport.stream(sourceStringsPageFetcherSplitIterator, false);
+        List<StringInfo> stringInfoToCheck = stream.collect(Collectors.toList());
+
+
+        logger.debug("done pulling strings for file {}, starting matching", fileUri);
+        Instant start = Instant.now();
+        Set<ThirdPartyTextUnitDTO> thirdPartyTextUnitDTOSet = convertStringInfoToDTO(stringInfoToCheck, fileUri);
```

Comment:
why do we have ThirdPartyTextUnitDTO and ThirdPartyTextUnitForBatchImport, could this be merged? Looks like we're doing conversion back to back.

Reply:
they are different objects and serve different purposes/have different fields, however since they are both based on ThirdPartyTextUnit they could be combined. I don't think its worth the rewrite though.